File: src/profile-manager.ts
import React, { useState, useEffect } from "react";
import { Box, Text, useInput } from "ink";
import Spinner from "ink-spinner";
import path from "path";
import os from "os";
import fs from "fs";

const getConfigPath = (): string => path.join(os.homedir(), CONFIG_FILE);
const getEvePath = (): string =>
  path.join(os.homedir(), "AppData", "Roaming", "EVE Online");
const getStateFilePath = (profileName: string): string =>
  path.join(getEvePath(), `state-${profileName}.json`);
const getLauncherStateFilePath = () => path.join(getEvePath(), "state.json");

const CONFIG_FILE = "eve-launch-manager.json";
const DEFAULT_CONFIG = {
  version: 1,
  activeProfile: "main",
  profiles: ["main"],
};

export interface Config {
  activeProfile: string;
  profiles: string[];
}

export class ProfileManager {
  public readonly config: Config;
  private onSetState?: (newState: Config) => void;

  constructor() {
    let config: Config = DEFAULT_CONFIG;
    if (fs.existsSync(getConfigPath())) {
      config = JSON.parse(fs.readFileSync(getConfigPath()).toString());
    } else {
      fs.writeFileSync(getConfigPath(), JSON.stringify(config, null, "\t"));
      fs.copyFileSync(getLauncherStateFilePath(), getStateFilePath("main"));
    }
    this.config = config;
  }

  useConfig() {
    const [curr, setCurr] = useState(this.config);
    this.onSetState = setCurr;

    return curr;
  }

  private updateConfig() {
    fs.writeFileSync(getConfigPath(), JSON.stringify(this.config, null, "\t"));
    if (this.onSetState) {
      this.onSetState(this.config);
    }
  }

  createProfile(name: string, basedOff?: string) {
    let contents = "{}";
    if (basedOff) {
      contents = fs.readFileSync(getStateFilePath(basedOff)).toString();
    }

    fs.writeFileSync(getStateFilePath(name), contents);

    this.config.profiles.push(name);
    this.updateConfig();
  }

  switchProfile(profileName: string) {
    // backup current profile first
    fs.copyFileSync(
      getLauncherStateFilePath(),
      getStateFilePath(this.config.activeProfile)
    );

    // copy the new state over state.json
    fs.copyFileSync(getStateFilePath(profileName), getLauncherStateFilePath());

    this.config.activeProfile = profileName;

    this.updateConfig();
  }
}

File: go-version-old/profile-manager.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
)

const (
	CONFIG_FILE = "eve-launch-manager.json"
)

// Config represents the configuration structure
type Config struct {
	ActiveProfile string   `json:"activeProfile"`
	Profiles      []string `json:"profiles"`
}

// DefaultConfig provides the default configuration
var DefaultConfig = Config{
	ActiveProfile: "main",
	Profiles:      []string{"main"},
}

// ProfileManager handles EVE Online profile management
type ProfileManager struct {
	Config Config
}

// getConfigPath returns the path to the config file
func getConfigPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, CONFIG_FILE)
}

// getEvePath returns the path to EVE Online directory
func getEvePath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, "AppData", "Roaming", "EVE Online")
}

// getStateFilePath returns the path to a profile's state file
func getStateFilePath(profileName string) string {
	return filepath.Join(getEvePath(), "state-"+profileName+".json")
}

// getLauncherStateFilePath returns the path to the launcher's state file
func getLauncherStateFilePath() string {
	return filepath.Join(getEvePath(), "state.json")
}

// NewProfileManager creates a new ProfileManager instance
func NewProfileManager() (*ProfileManager, error) {
	config := DefaultConfig

	configPath := getConfigPath()
	if _, err := os.Stat(configPath); err == nil {
		data, err := os.ReadFile(configPath)
		if err != nil {
			return nil, err
		}

		if err := json.Unmarshal(data, &config); err != nil {
			return nil, err
		}
	} else {
		// Create default config file
		data, err := json.MarshalIndent(config, "", "\t")
		if err != nil {
			return nil, err
		}

		if err := os.WriteFile(configPath, data, 0644); err != nil {
			return nil, err
		}

		// Copy initial state file
		sourceFile := getLauncherStateFilePath()
		destFile := getStateFilePath("main")
		if err := copyFile(sourceFile, destFile); err != nil {
			return nil, err
		}
	}

	return &ProfileManager{Config: config}, nil
}

// updateConfig writes the current configuration to disk
func (pm *ProfileManager) updateConfig() error {
	data, err := json.MarshalIndent(pm.Config, "", "\t")
	if err != nil {
		return err
	}
	return os.WriteFile(getConfigPath(), data, 0644)
}

// CreateProfile creates a new profile
func (pm *ProfileManager) CreateProfile(name string, basedOff string) error {
	var contents []byte
	var err error

	if basedOff != "" {
		contents, err = os.ReadFile(getStateFilePath(basedOff))
		if err != nil {
			return err
		}
	} else {
		contents = []byte("{}")
	}

	if err := os.WriteFile(getStateFilePath(name), contents, 0644); err != nil {
		return err
	}

	pm.Config.Profiles = append(pm.Config.Profiles, name)
	return pm.updateConfig()
}

// SwitchProfile switches to a different profile
func (pm *ProfileManager) SwitchProfile(profileName string) error {
	// Backup current profile
	if err := copyFile(getLauncherStateFilePath(), getStateFilePath(pm.Config.ActiveProfile)); err != nil {
		return err
	}

	// Copy new state over
	if err := copyFile(getStateFilePath(profileName), getLauncherStateFilePath()); err != nil {
		return err
	}

	pm.Config.ActiveProfile = profileName
	return pm.updateConfig()
}

// copyFile is a helper function to copy files
func copyFile(src, dst string) error {
	input, err := os.ReadFile(src)
	if err != nil {
		return err
	}

	return os.WriteFile(dst, input, 0644)
}
